File: cases/taint/c/callback-registration-negative/callback_registration_negative.c
struct Registry {
    void (*hooks[4])(int);
    int count;
};

int dfb_source(void) { // DFB-SOURCE: callback-registration-input
    return 1;
}

void dfb_sink(int value) {} // DFB-SINK: callback-registration-sink

void register_hook(struct Registry *registry, void (*hook)(int)) {
    registry->hooks[registry->count] = hook;
    registry->count = registry->count + 1;
}

void fire(struct Registry *registry, int value) { // DFB-WITNESS: callback-registration-fire
    for (int index = 0; index < registry->count; index++) {
        registry->hooks[index](value);
    }
}

void drop(int value) {
    dfb_sink(0);
}

void run(void) {
    struct Registry registry;
    registry.count = 0;
    register_hook(&registry, drop);
    fire(&registry, dfb_source());
}

File: cases/taint/kotlin/callback-registration-negative/CallbackRegistrationNegative.kt
package dataflowbench

object CallbackRegistrationNegative {
    class Registry {
        val hooks: MutableList<(String) -> Unit> = mutableListOf()

        fun register(hook: (String) -> Unit) {
            hooks.add(hook)
        }

        fun fire(value: String) { // DFB-WITNESS: callback-registration-fire
            for (hook in hooks) {
                hook(value)
            }
        }
    }

    fun dfb_source(): String { // DFB-SOURCE: callback-registration-input
        return "tainted"
    }

    fun dfb_sink(value: String) {} // DFB-SINK: callback-registration-sink

    fun drop(value: String) {
        dfb_sink("clean")
    }

    fun run() {
        val registry = Registry()
        registry.register(::drop)
        registry.fire(dfb_source())
    }
}
